### DISCUSSION TITLE
FLD File exporting slows down after several "for" iterations.

### DISCUSSION THREAD

**Discussion by C-Ehmke** [ansys/pyaedt#5284] [Q&A]
Hi guys,

Hope everyone is doing great.
I'm working on a script using PyAEDT to automatically export .fld files of a design with a lot of parametric variations (500 to 1000 variations). For this application, the .fld file needs to have a good density of points (about 130000).
I'm no expert in python, but the algorithm is working as intended (I know it could be better written, but for now it's what I need). In this script, I created a function that reads all solutions and uses a for loop to iterate each variation and export the .fld file using the hfss.post.export_field_file_on_grid() function.

Here is the code:
```
import pyaedt

def simulate_all():
    hfss.analyze_all()

def close_app():
    d.save_project() 
    d.release_desktop() 

def export_loss():
    grid_start_1 = [-122.4, -122.4, 0]
    grid_stop_1 = [122.4, 122.4, 30]
    grid_step_1 = [3.653, 3.653, 1]

    var1= hfss.available_variations.variations("Auto1 : LastAdaptive")
    cont = 0
    for item in var1:
        var = item
        setup = "Auto1 : LastAdaptive"
        path = "Volume_Loss_Density_case_"+str(cont)+".fld"
        hfss.post.export_field_file_on_grid("VolumeLossDensity",setup,var,path,'Cartesian', [0, 0, 0], grid_start_1,  grid_stop_1,  grid_step_1, intrinsics="2.45GHz")
        cont+=1

project_name = "ExemploParametrics.aedt"
aedt_version = "2024.2"
non_graphical = True
design_name = "HFSSDesign1"
d = pyaedt.launch_desktop(aedt_version, non_graphical=non_graphical, new_desktop_session=True)
hfss = pyaedt.Hfss(projectname=project_name,designname=design_name)

export_loss()
close_app()
```

The problem is that, in the beginning the exportation procedure is satisfactory and efficient, like 5 or 6 .fld files per minute. After a while, I noticed that the number of files exported per minute decreases, 1 or 2 per minute. For what I noticed, it gets worse as the loop goes on. This is something that surprised me because I thought it would be constant since this was the only thing running in my machine. I'm sending some images to illustrate and with the CPU and RAM monitor.

![cpu_while_exporting_1](https://github.com/user-attachments/assets/27967dc7-4c7a-4343-a190-2e52db002806)
![cpu_while_exporting_2_Git](https://github.com/user-attachments/assets/94f56cc9-89e3-4486-9f14-cdf2753aba29)

From what I read, maybe this could be a cache problem since the RAM consumption increases with time. Also, I've seen that it is possible to use memorization to make it more time efficient. Once again, I'm no python expert, so I'm probably missing something here.

Has anyone faced this issue before? How could I make it more efficient? Any suggestions would be very appreciated. 
Thank you in advance!

**Comment by gmalinve** [comment]
Hi @C-Ehmke,

Do you need a .fld file per variation? because if this is not the case you could avoid calling the method for every variation and simply pass all the variations at once.
You can call the method once and simply provide: hfss.available_variations.nominal_w_values_dict.

Hope this helps,

Giulia

**Reply by C-Ehmke** [reply]
Hello gmalinve,

Thank you for the reply!
I believe that it would be better to have one archive per variation, but I tried to call the hfss.post.export_field_file_on_grid with the hfss.available_variations.nominal_w_values_dict and it didn't work. Maybe I'm doing something wrong.

`   

    var = hfss.available_variations.nominal_w_values_dict
    hfss.post.export_field_file_on_grid("VolumeLossDensity", setup, var, path, 'Cartesian', [0, 0, 0], 
                                            grid_start_1, grid_stop_1, grid_step_1, intrinsics="2.45GHz")

`
Using that, I received the following error:
PyAEDT ERROR: AEDT GRPC API CALL ERROR on export_field_file_on_grid

It seems the hfss.available_variations.nominal_w_values_dict returns only the nominal values, and not all possible variations.
return -> {'Var': '-20mm', 'VarX': '-44mm'}

Best regards,

**Comment by gmalinve** [comment]
The _nominal_w_values_dict_ has to work because this is the default value used if no variation is provided.
You for loop is not actually correct because you ar looping through: _self.aedtapp.available_variations.variations()_ which return a list. In my dummy example: 
![image](https://github.com/user-attachments/assets/6aae6abb-5597-49ae-8a15-21bbfa5a026a)
in fact if I try to pass it to export_field_file_on_grid it fails. _variations_ has to be a dict.
So there should be something else wrong in your project. 
Would it be possible to share an example as a zip file?

**Reply by C-Ehmke** [reply]
Yeah, I understand what you're saying but I'm using a list and it works :)
Maybe it could be the pyaedt version that I used (0.7), but for what I checked  the method hasn't changed. Or it could be related to the "for" iteration method. In any case, the code that I sent previously works, exporting the fld files without the error you comment.
I asked to print the argument that I'm using for each iteration:

> PyAEDT INFO: pyaedt v0.7.4
> PyAEDT INFO: Python version 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:26:36) [MSC v.1937 64 bit (AMD64)]
> PyAEDT INFO: AEDT 2024.2.0 Build Date 2024-05-23 02:23:50
> Returning found desktop with PID 18312!
> PyAEDT INFO: Project ExemploParametrics has been opened.
> PyAEDT INFO: AEDT file load (threaded) time: 1.6994450092315674
> PyAEDT INFO: Aedt Objects correctly read
> PyAEDT INFO: aedt file load time 0.011728525161743164
> The argument used in the export_field on grid is:
> ['Var:=', ['-20mm'], 'VarX:=', ['-44mm']]
> PyAEDT INFO: Modeler3D class has been initialized! Elapsed time: 0m 0sec
> PyAEDT INFO: PostProcessor class has been initialized! Elapsed time: 0m 0sec
> PyAEDT INFO: Exporting VolumeLossDensity field. Be patient
> The argument used in the export_field on grid is:
> ['Var:=', ['-120mm'], 'VarX:=', ['-120mm']]
> PyAEDT INFO: Exporting VolumeLossDensity field. Be patient
> The argument used in the export_field on grid is:
> ['Var:=', ['70mm'], 'VarX:=', ['-120mm']]
> PyAEDT INFO: Exporting VolumeLossDensity field. Be patient

![image](https://github.com/user-attachments/assets/8b5c57fe-4a2f-483d-ab3c-4f8acab152a6)


I could send you the example I'm working on, but unfortunately it is bigger than 25MB. But using the code I sent and calling a HFSS example it should work just as well, just as long as the save field is enabled.


By the way, it's not clear to me how the export_field_file_on_grid method could handle multiple variations in just one export since this is not even possible in the software's field calculator. You need to select one variation and it doesn't accept multiple inputs.![image](https://github.com/user-attachments/assets/11b13722-b187-49f5-b07c-2f62144e9eca)



Regards,

**Comment by gmalinve** [comment]
@C-Ehmke is there a reason why you are using 0.7? if not please update to 0.11 and retry.
Please do let me know!

**Comment by gmalinve** [comment]
Hi @C-Ehmke,

Please upgrade to PyAedt 0.11.1 and simply try this:
`vars1 = hfss.available_variations.variations(setup_sweep="Setup1 : LastAdaptive", output_as_dict=True)`
this will return a list of dict that you can loop over as:
```
for item in vars1:
    var = item
    setup = "Setup1 : LastAdaptive"
    path = "Volume_Loss_Density_case_" + str(cont) + ".fld"
    file_path = hfss.post.export_field_file_on_grid(quantity="VolumeLossDensity",
                                                    solution=setup,
                                                    variations=var,
                                                    file_name=hfss.working_directory,
                                                    grid_type="Cartesian",
                                                    grid_stop=grid_stop,
                                                    grid_step=grid_step,
                                                    intrinsics="5GHz")
    cont += 1
```

Please do let me know if this works.

Kind regards,

Giulia

**Reply by C-Ehmke** [reply]
Hi @gmalinve 

Thank you for the suggestion. I updated pyaedt and now it is possible to use the dict as the argument.
Then, just to be clear, it is not possible to export all the Volume Loss Density in one single .fld file, right?

Nevertheless, the real problem persists. The exporting process is efficient in the beginning but as the loop goes on the process becomes less and less efficient. 

The RAM consumption increases and the CPU shows idle periods.
![image](https://github.com/user-attachments/assets/748d6d2f-4522-4eb8-9ec8-5ed5bc0e503c)

If there is any suggestion regarding this problem, I would appreciate it!

Regards,

**Comment by gmalinve** [comment]
Hi @C-Ehmke,

No it's not possible, even if you try manually and record the script you see that only the variations at the top is taken.
Regarding the unefficiency of the process I'm not sure it's a PyAEDT issue. What I could suggest is to maybe add at the end of each loop a sleep time of some seconds in order to allow time to AEDT to save the fields.
Another thing I could think of is to deactivate these options in AEDT:
![image](https://github.com/user-attachments/assets/78907f19-4cb7-4206-b7eb-d1942e0782d3)
and try to run it in non-graphical mode, plus deactivate the autosave.
Bear in mind that if you deactivate the autosave and you suddenly encounter any probem you can't recover your work.

Hope this helps,

Kind regards,

Giulia
